File: test.spritebuilder/Source/TDEngine/Utils/StackObjectAbstractFabric.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package Utils;

/**
 *
 * @author Константин
 */
import Utils.StackObject.StackObjectAlreadyEngaged;
import java.util.*;

/**
 * Created by IntelliJ IDEA.
 * User: K.Vereschagin
 * Date: 25.10.12
 * Time: 16:04
 * To change this template use File | Settings | File Templates.
 */
public abstract class StackObjectAbstractFabric  {
      public static int countOfStartObject = 200;
    Dictionary<Integer,ArrayList<StackObject>> objects;
         
         public StackObjectAbstractFabric( )
    {
            List<Integer> types=  new ArrayList<Integer>();
            types.add(0);   
            objects = new Hashtable<Integer, ArrayList<StackObject>> ();
           initializeStartObject(types);
    }
    public StackObjectAbstractFabric(List<Integer> types)
    {
        objects = new Hashtable<Integer, ArrayList<StackObject>> ();
        initializeStartObject(types);
    }

    protected void initializeStartObject(List<Integer> types)
    {
         for (int i = 0; i< types.size(); i++)
         {
             int currentID = types.get(i);
             ArrayList<StackObject> currentList =   new ArrayList<StackObject>();
             for(int j =0;j<countOfStartObject;j++)
                 currentList.add(getNewObject(currentID));
             objects.put(currentID,   currentList   );
         }
    }

    protected StackObject getObject(Integer idObject) throws StackObjectAlreadyEngaged
    {
        StackObject obj =  findFreeObject(objects.get(idObject));
        if(obj == null)
            obj = getNewObject(idObject);
        obj.UseStackObject();
        return obj;
        
    }

    protected  abstract StackObject getNewObject(Integer idObject);


    StackObject  findFreeObject(ArrayList<StackObject> objectsType)
    {
        for(int i = 0; i < objectsType.size();i++)
        {
           if( objectsType.get(i).isFree())
               return objectsType.get(i);
        
        }

        return null;
    }


}

File: TowerDefence.spritebuilder/Utils/StackObjectAbstractFabric.swift
//
//  StackObjectAbstractFabric.swift
//  TowerDefence
//
//  Created by koveres on 02.12.14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

import Foundation


public  class StackObjectAbstractFabric  {
   
    var objects:[CInt:[StackObject]];
    
    
  public  init(intlist:[CInt] )
    { 
        objects = [CInt:[StackObject]]();
        initializeStartObject(intlist);
    }
    
    
    public func StackObjectAbstractFabric( types : [CInt])
    {
        objects = [CInt:[StackObject]]();
        initializeStartObject(types);
    }
    
    
    func initializeStartObject(types:[CInt])
    {
        
        for currentID in types {
            var currentList:[StackObject] = [StackObject]();
           
        }
        
    }
        
        
    
    
    func  getObject( idObject: CInt) -> (StackObject)
    {
        var obj:StackObject? =  findFreeObject(objects[idObject]);
        if(obj == nil)
        {
            obj = getNewObject(idObject,grObj: EngineOptions.instance_.value!.graphObjCreator.getGrObjectUnit(idObject)());
        }

        obj!.UseStackObject();
        
        return obj!;
    
    }
    
        
    func     getNewObject( idTypeObject: CInt, grObj:AnyObject) -> (StackObject){fatalError("Calling abstract function.");}
    
    
        
        
    func  findFreeObject( objectsType :[StackObject]?) -> (StackObject?)
    {
        
        if(objectsType == nil) {return nil;}
        
        
        for object in objectsType!
        {
            if( object.isFree()){
                return object;
            }
        
        }
        
        return nil;
    }
    
    
}

